### PR TITLE
Mime Powers (Vow + Invisible Wall)

### PR DESCRIPTION
**Pull request by Elijahrane**
## About the PR 

**Screenshots**

**Changelog**

:cl: Rane
- add: Mimes can now create invisible walls using their secret knowledge of their craft.
- remove: Mimes can no longer speak or scream without revoking their Vow of Silence

### CODE REVIEW

File: Content.Server/Abilities/Mime/InvisibleWallComponent.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+namespace Content.Server.Abilities.Mime
+{
+    // Tracks invisible wall despawning
+    [RegisterComponent]
+    public sealed class InvisibleWallComponent : Component
+    {
+        [DataField("accumulator")]
+        public float Accumulator = 0f;
+
+        public float DespawnTime = 30f;
```
Comment (mirrorcult) [inline Content.Server/Abilities/Mime/InvisibleWallComponent.cs:10]: make this a timespan and datafield

File: Content.Server/Abilities/Mime/MimePowersSystem.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            {
+                invisWall.Accumulator += frameTime;
+                if (invisWall.Accumulator < invisWall.DespawnTime)
+                {
+                    continue;
+                }
+                EntityManager.QueueDeleteEntity(invisWall.Owner);
+            }
+            /// Queue to track whether mimes can retake vows yet
+            foreach (var (vowbreaker, mime) in EntityQuery<VowbreakerComponent, MimePowersComponent>())
```
Comment (mirrorcult) [inline Content.Server/Abilities/Mime/MimePowersSystem.cs:49]: if the mime powers component doesn't get removed when the vow is broken, then the vowbreaker comp shouldnt be separate at all imo.

Comment (mirrorcult) [reply Content.Server/Abilities/Mime/MimePowersSystem.cs:49]: either:
1) turn the mime actions into action prototypes, remove `MimePowersComponent` (and add vowbreaker) when the vow is broken, remove the `Enabled` field from `MimePowersComponent` and readd `MimePowersComponent` when they take the vow again
2) merge vowbreaker fields into mime powers, remove the queue stuff

File: Content.Server/Abilities/Mime/MimePowersSystem.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            /// Check there are no walls or mobs there
+            foreach (var entity in coords.GetEntitiesInTile())
+            {
+                if (_tagSystem.HasTag(entity, "Wall") || (HasComp<MobStateComponent>(entity) && entity != uid) ||
+                    (TryComp<DoorComponent>(entity, out var door) && door.State == DoorState.Closed))
+                {
+                    _popupSystem.PopupEntity(Loc.GetString("mime-invisible-wall-failed"), uid, Filter.Entities(uid));
+                    return;
+                }
+            }
```
Comment (mirrorcult) [inline Content.Server/Abilities/Mime/MimePowersSystem.cs:97]: feels like it should just check
1) that there are no mobs
2) that there are no entities with an impassible fixture

File: Content.Server/Abilities/Mime/MimePowersSystem.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            _popupSystem.PopupEntity(Loc.GetString("mime-invisible-wall-popup", ("mime", uid)), uid, Filter.Pvs(uid));
+            /// Make sure we set the invisible wall to despawn properly
+            var wall = EntityManager.SpawnEntity(component.WallPrototype, coords);
+            var invisWall = EnsureComp<InvisibleWallComponent>(wall);
```
Comment (mirrorcult) [inline Content.Server/Abilities/Mime/MimePowersSystem.cs:101]:
```suggestion
            EnsureComp<InvisibleWallComponent>(wall);
```

File: Content.Server/Alert/Click/BreakVow.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    public sealed class BreakVow : IAlertClick
+    {
+        public void AlertClicked(EntityUid player)
+        {
+           if (IoCManager.Resolve<IEntityManager>().TryGetComponent<MimePowersComponent?>(player, out var mimePowers))
+           {
+                EntitySystem.Get<MimePowersSystem>().BreakVow(player, mimePowers);
+           }
+        }
+    }
```
Comment (mirrorcult) [inline Content.Server/Alert/Click/BreakVow.cs:19]: breaking/retaking the vow should really be actions IMO. also can make them yaml prototype actions

Comment (mirrorcult) [reply Content.Server/Alert/Click/BreakVow.cs:19]: actuaaaaally im not sure. having it be an alert click is fine, i think? can get a second opinion on that

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Content.Server/Abilities/Mime/InvisibleWallComponent.cs (hunk 1 above)
Comment (mirrorcult) [reply Content.Server/Abilities/Mime/InvisibleWallComponent.cs:10]: despawntime specifically, nto accumulator

File: Content.Server/Alert/Click/BreakVow.cs (hunk 5 above)
Comment (Elijahrane) [reply Content.Server/Alert/Click/BreakVow.cs:19]: Chose alert clicks because those are harder to do accidentally

Comment (mirrorcult) [reply Content.Server/Alert/Click/BreakVow.cs:19]: also true

**Comment by Elijahrane** [comment]
Reviews addressed

**Review by mirrorcult — APPROVED** [review]

**Pull request merged**